Support for tagging user lists

samuli opened the pull request:
This PR adds basic support for assigning tags to user lists. User list tags are saved in the same manner as when assigning tags to resources, i.e. by updating entries in tables `tags` and `resource_tags`. User list tags are separated from resource tags by allowing `resource_id` field in `resource_tags` to be NULL.

Current functionality should not be affected. User list tagging is disabled by default and can be enabled via a new setting. When enabled, tags can be assigned to the list in the edit-mode and the assigned tags are displayed on the mylist-page. Enabling user list tags does not affect Tag-search and Tag-browsing, they are restricted to cover only resource tags.  

Our use case is to allow retrieving user lists by a given tag, for example by adding a `listSearch`method to `tags` table, similar to `resourceSearch`. This could be wired to the `Tag` controller, maybe by passing in a parameter that restricts the search to user lists. I think we can build this to our own code base with the help of this PR. Another option would be to extend VuFinds' current tag-search to cover user list tags as well.

What are your ideas about this?

TODO
- [x] Open JIRA ticket with reminder to write Mink tests to cover new functionality ([VUFIND-1416](https://vufind.org/jira/browse/VUFIND-1416))
- [x] Add PostgreSQL migration
- [x] Test with PostgreSQL
- [x] When merging, update changelog with minor BC breaks (e.g. UserList row constructor signature change)
- [x] When merging, update database changelog and improve documentation of resource_tags table

demiankatz commented:
Thanks, @samuli, this seems like a reasonable idea. Right now I'm trying to focus my attention on finishing up the work in progress for the 7.0 release, so I may not be able to review this immediately, but I'll leave some more detailed comments as soon as time permits.

demiankatz commented on module/VuFind/src/VuFind/Controller/AbstractBase.php:
```diff
@@ -638,6 +638,18 @@ protected function tagsEnabled()
+     *
+     * @return bool
+     */
+    protected function listTagsEnabled()
```
Since this is only used in the MyResearchController, does it need to be placed at the AbstractBase level?

demiankatz commented on module/VuFind/src/VuFind/Db/Row/User.php:
```diff
@@ -275,28 +275,34 @@ public function getTags($resourceId = null, $listId = null, $source = null)
      * @return string
      */
-    public function getTagString($resourceId = null, $listId = null, $source = null)
+    public function getTagString($tags)
```
Rather than modifying getTagString(), maybe it would be better to refactor this, so there's a new formatTagString method that gets used by getTagString and can also be called separately. That would introduce fewer backward compatibility breaks. We can always deprecate getTagString() for future removal if it's no longer useful.

demiankatz commented on module/VuFind/src/VuFind/Db/Table/ResourceTags.php:
```diff
@@ -237,10 +237,16 @@ public function destroyLinks($resource, $user, $list = null, $tag = null)
             if (null !== $resource) {
-                if (!is_array($resource)) {
-                    $resource = [$resource];
+                if ($resource === 'listTag') {
```
Would it be better to have a separate method rather than using a special case here? This is one of several things that makes me wonder if we'd be better off having a separate table for this functionality rather than overloading user_resource/resource_tags.

demiankatz replied:
I still don't like the use of a magic string here. Would it be better to create separate destroyListLinks / destroyResourceLinks methods? If that's too redundant/inefficient, maybe we could at least replace the hard-coded string with a class constant to make it more clear that it has a very specific meaning.

demiankatz commented on module/VuFind/sql/mysql.sql:
```diff
@@ -260,7 +260,7 @@ CREATE TABLE `user_list` (
   `id` int(11) NOT NULL AUTO_INCREMENT,
   `user_id` int(11) NOT NULL,
-  `resource_id` int(11) NOT NULL,
+  `resource_id` int(11) DEFAULT NULL,
```
I'm confused; why are you changing the constraint on user_resource rather than on resource_tags?

Do we really need to overload existing tables, or could we create a new user_list_tags table to capture this information in a clearer way?

samuli commented:
@demiankatz My motivation for re-using existing tables was to prevent duplicating the tag handling logic (deleting, adding etc) for list tags. But I agree that this might be too confusing... so maybe we should proceed with a new linking table (like resource_tags but without `resource_id`) and use this via a new table class (possibly extended from `ResourceTags`). Perhaps it's also better to create a new table for the actual list tags? Otherwise `ResourceTags` would need to check also this new table which would again introduce an unnecessary coupling between the two tables.

demiankatz commented:
@samuli, I think what this conversation may be pointing us toward is the realization that the tag handling logic is not currently in the best place in the code. I've never been entirely comfortable with some of the logic in the table classes, because its location is sometimes a bit arbitrary. I wonder if we'd be better off moving some of it out of the low-level table classes and into a higher-level tag service.

I'm also a bit uncomfortable with the way record drivers are used for adding/deleting tags (and other database-related activities). This seems to be overloading them to do more than their core purpose of retrieving data about records. 

At the risk of some significant scope creep here, I wonder if this would all become easier and more clear if we did some refactoring to create "addTags" and "deleteTags" methods on the VuFind\Tag service, and rearranged code to decouple this functionality from the record drivers and low-level table classes. The VuFind\Tag service would be responsible for manipulating the table classes, and we would access that service directly from controllers instead of calling it through the record driver. I think this would make everything more appropriately decoupled and centralized, make testing easier, and also allow you to avoid creating two different tables for list vs. resource tags (since the shared service could cleanly handle orphan removal from a single table, without having to duplicate that logic in two lower-level link table classes).

Does that make any sense?

Whether or not it does, I wonder if it would make sense to start a separate PR to do some refactoring. If you agree with the idea, perhaps you could take the lead; if you're not entirely sure what I'm talking about, I could put together an initial PR to demonstrate, which might clarify my ideas. (We could always throw it away if it turned out not to be useful).

Obviously, if we do this degree of refactoring, we won't be able to merge it until VuFind 8, since it will introduce some backward compatibility breaks. I think it may be worthwhile if it cleans up and clarifies the architecture, but if you are in a hurry to implement this functionality sooner, we may need to figure out a different compromise for 7.1.

What do you think?

samuli commented:
@demiankatz I agree that moving the logic to Tag-service is a good idea. We were planning to include list tagging in our next release, in about three weeks. However, it's not required that the feature is bundled into a VuFind-release, but we would need the feature to be merged into VuFind's master so that we can cherry-pick from there. Whether or not we are going to have a temporary solution, I suppose it's best to decide the data-model so that we don't need to migrate data later. 

- If we want to use `tags` table for both tag types, should we also share `resource_tags` (and use `resource_id` to separate between the two types)? Or is it more clear to create a new table `list_tags`?
- Tag-service methods could be parametrized to know about the passed tag type (resource / list) and act accodingly.

demiankatz commented:
Thanks, @samuli. It does seem that simply overloading the resource_tags table is the path of least resistance, since it allows us to create a non-backward-breaking implementation now and deal with refactoring code later... though as noted in one of my code-level comments, it looks like this PR is currently changing the constraint on resource_id in the wrong table.

I think it would make the data more clear to have a separate list_tags table, but it might at the same time make the code less clear by increasing the number of classes and introducing redundant logic. As it is, we're already overloading the resource_tags table to a certain extent by using it for both tagging resources in lists and also tagging resources directly with a null list_id. This whole model feels like it may not be normalized in the best possible way (for example, the "tag on record" scenario doesn't really need a list_id, and the "tag on list entry" scenario doesn't really need a user_id, at least unless we introduce shared lists in the future). I think if we were to create a separate list_tags table, we might also want to consider breaking apart the other two use cases into separate tables.

Bottom line: I think I've convinced myself that the database solution you originally proposed is probably the most pragmatic approach, even if it may not be the best in an ideal world. We should revisit all of this in VuFind 8 (maybe it's worth opening a JIRA ticket to track the ideas), but I think we can go forward with this in its current form for 7.1. If we refactor the code to the tag service, we make it easier to restructure the data later if there's a compelling reason to, but it's entirely possible that the existing model will work just fine and will not need to be changed. We should just be sure to clearly document how the table is used, since at a glance it is not immediately obvious. (This documentation could exist in the [wiki](https://vufind.org/wiki/development:architecture:database_schema#resource_tags) and/or comments in the tag service code, when that is eventually written).

I'm going to be on holiday for most of the next two weeks, so it may be difficult to get this fully reviewed and merged in time for your planned release date... but I will do my best! I suppose at this point the next step is for you to look at my existing comments, since it sounds like we can get away without major low-level changes for the moment.

samuli commented:
@demiankatz I didn't realize that `resource_tags` handles tagging resources both within a list-context and outside a list... This is quite flexible, and I think tagging lists by setting `resource_id` to NULL would extend this pattern without being too hacky. So then we would have the following cases in `resource_tags`:
| resource_id | list_id | Use case |
| --- | --- | --- |
| X | X | Resource tag within a list |
| X | null | Resource tag outside a list |
| null | X | User list tag |

- Continuing to having `user_id` optionally null is worthwhile since it allows us to support private and public tags in the future. We already have public and private lists, but upcoming list tags are going to be used only for public lists (at least for now). 
- Storing all tags in a shared tag-table is nice if we want to extend search-by-tag to include also other things than records (lists are currently discarded from search results). In addition to lists, tagging could be extended (for example) to handle "sub-resources" belonging to a single record (links, images etc).

> though as noted in one of my code-level comments, it looks like this PR is currently changing the constraint on resource_id in the wrong table.

Yes, this was a mistake. Edited to change the constraint on `resource_tags`.

Travis fails because the factory for VuFind\Tags doesn't work in the context of `UserListTest` even though the test properly uses `UserListFactory`. Any idea how this could be fixed?

demiankatz commented:
Thanks, @samuli, this all makes sense!

Regarding the failing test, the service manager is set up differently in test mode than it is in normal VuFind operations; the test classes have to define all of the services in the container. This is something I would like to simplify and clean up at some point, but for now, the easiest solution would be to set up the VuFind\Tags service in the [DbTestCase](https://github.com/vufind-org/vufind/blob/dev/module/VuFind/src/VuFindTest/Unit/DbTestCase.php#L102) so it is defined when the factory tries to use it.

Please let me know if you need more help, and when you're ready for another review.

demiankatz commented on module/VuFind/sql/pgsql.sql:
```diff
@@ -47,7 +47,7 @@ DROP TABLE IF EXISTS "resource_tags";
 CREATE TABLE resource_tags (
 id SERIAL,
-resource_id int NOT NULL DEFAULT '0',
+resource_id int DEFAULT NULL,
```
We should also add a PostgreSQL migration for this change; I've added a TODO list to the PR so we don't forget to do this. I'm perfectly happy to hold off on that detail until closer to the end of the project, so we don't have to revise it if we decide to make any further changes in the meantime.

demiankatz commented on module/VuFind/src/VuFind/Controller/MyResearchController.php:
```diff
@@ -909,8 +909,22 @@ public function mylistAction()
+
+            if ($this->listTagsEnabled()) {
+                $list = $results->getListObject();
+                $user = $this->getUser();
```
Does the `$user` check here mean that tags are not visible on public lists unless the owner is logged in? Is that intentional? Is there a reason not to use `$list->getListTags()` here instead?

samuli replied:
This was not intentional. Edited to use `$list->getListTags` so that public list tags are displayed also for non-owners.

demiankatz commented on module/VuFind/src/VuFind/Db/Row/UserList.php:
```diff
@@ -79,7 +89,7 @@ public function editAllowed($user)
 
     /**
-     * Get an array of tags associated with this list.
+     * Get an array of resource tags associated with this list.
```
Maybe it's worth renaming the method to match this comment. You could keep a deprecated getTags() method as an alias for getResourceTags() for backward compatibility. This might make other code more clear.

demiankatz commented on module/VuFind/src/VuFind/Db/Table/ResourceTags.php:
```diff
@@ -237,10 +237,18 @@ public function destroyLinks($resource, $user, $list = null, $tag = null)
+                    // (resource_id is NULL for list tags).
+                    $select->where->isNull('resource_id');
+                } else {
+                    if (!is_array($resource)) {
```
While you're changing this code anyway, you could optimize it to:

`$select->where->in('resource_id', (array)$resource);`

and get rid of the preceding `if` statement.

samuli commented:
@demiankatz Thank you for your helpful comments. I've separated `destroyLinks` into `destroyResourceLinks` and `destroyListLinks` and it's much clearer now.

samuli commented:
@demian I made a small change to saving list edits: tags are updated only when tagging is enabled in the settings (https://github.com/vufind-org/vufind/pull/1645/commits/77daa35500091b5e32585f16dcf3638ddf8b629e). This way the tags will have to be explicitly deleted.

demiankatz commented on themes/bootstrap3/templates/myresearch/mylist.phtml:
```diff
@@ -71,6 +71,9 @@
     <p><?=$this->escapeHtml($list->description)?></p>
   <?php endif; ?>
+  <?php if (!empty($listTags)): ?>
+    <div><?=$this->transEsc('Your Tags')?>: <?=implode(', ', array_map([$this, 'escapeHtml'], $listTags))?></div>
```
I wonder if this label should simply be "Tags" since in public mode, the tags don't belong to the viewing user.

demiankatz commented on module/VuFind/src/VuFind/Db/Row/UserList.php:
```diff
@@ -93,6 +103,32 @@ public function getTags()
 
+    /**
+     * Get an array of resource tags associated with this list
+     * (deprecated, use getResourceTags).
```
You can use a `@deprecated` annotation in the comment, which might improve the PHPDoc (and make it easier for me to find and remove this in the future).

demiankatz commented on module/VuFind/src/VuFind/Db/Table/ResourceTags.php:
```diff
@@ -232,15 +232,12 @@ public function getStatistics($extended = false)
      * @return void
      */
-    public function destroyLinks($resource, $user, $list = null, $tag = null)
+    public function destroyResourceLinks($resource, $user, $list = null, $tag = null)
```
Can you add back a deprecated `destroyLinks` method for backward compatibility? It can simply proxy `destroyResourceLinks`, and we can remove it in the next major release.

demiankatz commented on module/VuFind/src/VuFind/Db/Table/Tags.php:
```diff
@@ -156,6 +158,7 @@ public function resourceSearch($q, $source = null, $sort = null,
             } else {
                 $select->where->equalTo('tags.tag', $q);
             }
+            $select->where->isNotNull('rt.resource_id');
```
Should this get the standard "Discard tags assigned to a user list." comment too?

samuli commented:
@demiankatz :
> If you have a setup for Mink testing, I'd certainly appreciate help on that

Sorry, unfortunately I don't... 

> what is the purpose of this functionality at this time?

We're going to implement a widget that displays user lists filtered by selected tags. The use case is to allow one to group lists belonging to a common theme together and make it easier for users  to find and browse them. For now we have grouped lists together by linking to the lists manually, but this is of course quite time consuming. 

The widget is going to be embedded into a content page and will be implemented as a view helper. The backend needs a new method that receives a list of tags (IDs or strings) and returns lists, maybe based on `Tags::resourceSearch`. I have planned to implement this directly into our code base, but we could make a version for VuFind's theme if it would be useful? Or contribute just the backend part (athough there wouldn't be any use for it without the presentation level/widget)?

Our widget is going to look like this:

![image](https://user-images.githubusercontent.com/3889/89651393-884e9200-d8cc-11ea-96ee-70c173d6b9e7.png)
- Selectable tags are pre-configured via view helper parameters and the widget doesn't allow one to input a tag freely. 
- The list preview image is not uploaded via the list-edit form. It has to be manually put into a theme directory (at least for now).

demiankatz commented:
@samuli, this looks good and makes sense!

Regarding the display widget, I wonder if the best way to integrate this with the VuFind core would be to add a tag-related setting to the ListItems ChannelProvider (https://github.com/vufind-org/vufind/blob/dev/module/VuFind/src/VuFind/ChannelProvider/ListItems.php). I don't think it would be too complicated to add, and then we'd have a built-in demonstration of the functionality.

Of course, this does raise one other issue: if you're using tags for administrative purposes, does that create a situation where tag editing needs to be limited with a permission setting, so only administrative users can edit tags? Otherwise, there's a risk of a malicious user tagging a list and invading your curated territory. Do we need better controls on this functionality?

samuli commented:
@demiankatz 

> I wonder if the best way to integrate this with the VuFind core would be to add a tag-related setting to the ListItems ChannelProvider

> if you're using tags for administrative purposes, does that create a situation where tag editing needs to be limited with a permission setting, so only administrative users can edit tags? 

Yes, I think this could work . I've added a draft-proposal, which still needs some work:
- lower-casing tag names in db-query
- the sort order doesn't seem to have an effect in the UI although the output from `ListItems` is correctly ordered...

Please let me know what you think of it. 

We are going to let list owners add tags freely but restrict the displayed lists before sending them to the widget (`ListItems::getListsByCriteria`). We'll also need to restrict the results to public lists only. This should not be possible to override via a configuration file. Maybe this restriction could be hardcoded tohg `getListsByCriteria`? Or then via a new more specific `PublicListItems` provider?

demiankatz commented on config/vufind/channels.ini:
```diff
@@ -128,13 +128,18 @@ maxValuesToSuggestPerField = 2
 ; ID specified, and these IDs will be displayed first.
 ;ids[] = 1
+;tags[] = tag1
+;userIds = 1
```
I think this should be `userIds[]`.

demiankatz commented:
Thanks, @samuli, this looks like a good start!

Regarding the sorting problem, is it possible that clearing your cache will help? I seem to recall that the channels code does some caching, and I wonder if the cache does not take the sort setting into account. (I could be wrong -- but worth trying before doing more involved troubleshooting/digging).

I also notice that this PR has a conflict with master for some reason. I'm not sure how that happened but please let me know if you have any trouble resolving it!

samuli commented:
@demiankatz 

I've extended `ListItems::getLists` to handle filtering based on id, tag and userId. It should work in a backward-compatible way. Filtering uses AND/OR depending on the field: `tag` uses AND, `id` & `userId` use OR. This makes sense when displaying the results in the widget, where selecting tags should narrow the results.

I simplified the previous attempt by removing the `sort` configuration option. The results are sorted in the same way as now, i.e. in the order configured by the id-list, possibly followed by public lists. 

> We'll also need to restrict the results to public lists only. This should not be possible to override via a configuration file

On a closer look, the provider already restricts the results to public lists. The same logic should continue to work with this new version. 

> Regarding the sorting problem, is it possible that clearing your cache will help?

Yes this helped and I also had an error in my SQL-query...

Regarding the time schedule, if you think channel provider changes will delay merging, I could separate them into a new PR?

demiankatz commented on module/VuFind/src/VuFind/Db/Table/ResourceTags.php:
```diff
@@ -186,6 +186,84 @@ public function getResourcesForTag($tag, $userId, $listId = null)
+        $tag, $listId = null, $userId = null, $publicOnly = true
+    ) {
+        $tag = $tag ? (array)$tag : null;
+        $listid = $listId ? (array)$listId : null;
```
There's a case problem here -- the first `$listid` should be `$listId`.

demiankatz commented on module/VuFind/src/VuFind/Db/Table/ResourceTags.php:
```diff
@@ -186,6 +186,84 @@ public function getResourcesForTag($tag, $userId, $listId = null)
+        $callback = function ($select) use (
+            $tag, $userId, $listId, $publicOnly
+        ) {
+            $select->columns(['list_id' => 'list_id', '*']);
```
Is there a reason to explicitly specify `list_id` here if you're also specifying `*`? It might also be slightly cleaner to use the `\Laminas\Db\Sql\Select::SQL_STAR` constant.

samuli replied:
No reason, I took the pattern from `getResourcesForTag`. Fixed and edited to use `SQL_STAR`.

demiankatz replied:
Yes, I see that I didn't follow my own advice there; I've opened https://github.com/vufind-org/vufind/pull/1690 to clean that up. :-)

demiankatz commented on module/VuFind/src/VuFind/ChannelProvider/ListItems.php:
```diff
@@ -178,49 +210,61 @@ protected function buildListChannels($channelToken)
+            $callback = function ($select) use ($resultIds) {
                 $select->where->equalTo('public', 1);
-                foreach ($ids as $id) {
+                foreach ($resultIds as $id) {
```
I realize this `foreach` existed in the original code, but (unless I'm missing something) I think it would be simpler to use a `$select->where->notIn('id', $resultIds)` clause instead.

demiankatz commented on module/VuFind/src/VuFind/ChannelProvider/ListItems.php:
```diff
@@ -164,9 +192,13 @@ public function getFromSearch(Results $results, $channelToken = null)
+            $this->initialListsToDisplay
+        );
         $channels = [];
-        $lists = $channelToken
```
In the original code, if we received a channel token, we would short circuit the whole lookup process and return only that channel. This is needed for AJAX loading specific channels from the drop-down list. It looks like in the new code, the channel token is getting passed to `getLists` along with all the other standard parameters; I think this may result in unexpected/undesired results. Maybe it's okay and I'm just not reading the code closely enough -- but wanted to be sure you understood the context of the channel token, since it may not be obvious at first glance.

samuli replied:
I tested this quickly in the UI and couldn't find unexpected results, although I might miss something. Overriding $this->ids with channel token (when received) and passing the other filtering parameters should not matter since the list for "Add more channels like this" is also filtered using the same parameters?

Although it should produce the same results, we could refactor `buildListChannel` to ignore filtering parameters when handling channel token:
```
if ($channelToken) {
     $lists = $this->getLists(
         null, [$channelToken], null, $this->initialListsToDisplay
     );
} else {
    $lists = $this->getLists(
        $this->tags,
        $this->ids,
        $this->userIds,
        $this->initialListsToDisplay
    );
}
```

demiankatz replied:
My concern is that, if we have a channel token, we know exactly which list ID we need to look up, and we want only that list and not anything else. With the current logic, for example, it looks to me like if we have the displayPublicLists setting turned on, we could potentially retrieve and add a whole bunch of extra public lists that we don't actually need. Additionally, generating the whole complicated SQL query to look up lists just to fetch a known ID is probably going to negatively impact performance. I think it might be worth bringing back the pre-existing `getListsById()` method to cover this special case.

samuli replied:
I reverted the handling of channelToken and ids to their current version. If tags are used (possibly together with ids), the new `getListsByTagAndId` is used. I agree that it's more clear and safe this way.

demiankatz commented on module/VuFind/src/VuFind/Db/Table/ResourceTags.php:
```diff
@@ -186,6 +186,84 @@ public function getResourcesForTag($tag, $userId, $listId = null)
+            if ($tag) {
+                // Use AND operator for tags
+                $select->having->literal(
+                    'COUNT(resource_tags.list_id) = ?',
```
Would it make more sense to `count(distinct(resource_tags.tag_id))` here? I'm concerned that counting the list_id might work now but might break in future if we extended the way we used the database (for example, if multiple users could add tags to a list).

samuli replied:
Good point, supporting tags also from non-list-owners (and filtering based on tag-creator) could be usable in the future. I tried `count(distinct(resource_tags.tag_id))` but it doesn't seem to work in a similar way. But could we achieve the same by adding an `where->in('resource_tags.user_id', $userId)` ? Btw: I accidentally used this in the previous version, although $userId should determine list-creator.... fixed now:
https://github.com/vufind-org/vufind/pull/1645/files#diff-40d3d2075f2211da8a41e54053a77d23R232

demiankatz replied:
Just to be sure I understand what you're trying to do here: the idea is that we want to AND together tag filters, rather than ORing them. When we say `$select->where->and->in('t.tag', $tag);`, we're bringing back all lists that have ANY of the requested tags... but when we `GROUP` by list_id, we want to apply a `HAVING` clause that will filter to only the lists that have ALL of the tags. If we can assume that each tag will be applied to each list only once, then we can filter on the groups that have a number of entries equal to the number of tags... but my concern is that this breaks down if there are scenarios where the same tag can be applied to a list multiple times by multiple users. To ensure that all requested tags are present, we would have to count distinct values, not just total number of rows. Some of the reading I've done suggests that this might work better if you add the count of distinct values as a selected value in the `columns()`, and then use the name of that value in the `HAVING` clause.

Does that help/make sense, or am I off base here?

samuli replied:
Yes, we want to AND together tags and OR ids. I revised the query to count by distinct tag id and use this in the HAVING clause. I also restricted the query to tags assigned by the list owner. Now the query should continue to work in the same way even if non-owners are allowed to tag lists in the future (which I think is a good thing to take into consideration).

About ANDing tags: this makes sense for our use-case, but should this be configurable? `ResourceTags::getListsForTag` could be passed the operator as a parameter and if OR is used, `GROUP BY`, `HAVING` and `COUNT(DISTINCT` could be dropped from the query.

demiankatz replied:
Thanks for the progress on this! It seems like it wouldn't hurt to add a "tagsOperator" setting to the .ini file and support it here by adding another parameter to the method. However, if this is more work than you want to spend time on, I think we can keep it as-is and work on that later as needed. I'll leave it up to you; just let me know your preference so I know whether or not to wait for more progress before doing a (hopefully) final round of testing.

samuli replied:
I added `tagsoperator`. We can leave it out if it makes things too complicated, but I think it could be useful.

samuli commented:
@demiankatz 

> What's your deadline for getting this all merged?

We're hoping to get list-tagging merged by the beginning of next week. Our next release is on Thursday next week and this would leave use some time for making sure everything works with our UI.

ChannelProvider changes can be delayed without problems, and I can leave them out of this PR if needed.

demiankatz commented:
Thanks, @samuli, I think we can get this sorted out by the end of this week, one way or another. Perhaps with a little more conversation tomorrow, we can get the channel provider finalized and be able to merge everything. If not, I can merge the basic underlying functionality to the dev branch on Friday and leave this PR open to finish up the remaining channel provider work in the near future.

demiankatz commented:
@samuli, I took care of the PostgreSQL migration and testing this morning, so I think everything is ready to go except for the channel provider. If you don't have time to make more progress there, I'll merge the rest either later today or over the weekend, depending on available time. Please note that I will be completely unavailable on Monday because I will be driving all day to return from my vacation; apologies in advance for any inconvenience caused by that!

demiankatz commented:
@samuli, I've merged the basic functionality from here as https://github.com/vufind-org/vufind/commit/2b05cef5c7f3041444ad415935f482e8b3cacfee, so now only the channel code remains in this PR. Please let me know if you encounter any problems!

samuli commented:
@demiankatz Thanks for your time and effort. We'll cherry-pick the merged functionality and I'll continue with the channel provider changes from here as soon as possible.

samuli commented:
@demiankatz I updated the code according to the comments above. In addition I also dropped $userId filtering, since filtering by id and tag is enough for our use case.

demiankatz commented:
Thanks, @samuli, I just replied to the last open thread, and when I hear your reply to that, I'll do a final round of review and hopefully get this merged. As always, I appreciate all the work you've done on this!

samuli commented:
@demiankatz This is ready for testing.

demiankatz commented:
Thanks, @samuli; I ran out of time today but should be able to work on this first thing tomorrow!

samuli commented:
@demiankatz Good catch on the spaceship operator... I tested the changes and everything seems to work as expected.

demiankatz commented:
Thanks for the testing! When reviewing the diffs this morning, I noticed that at the end of my refactoring, I had left some meaningless changes in `buildListChannels()`, so I put it back the way it was to make the diffs slightly less confusing. I'm merging now!

demiankatz commented:
I accidentally bumped the enter key and submitted a blank commit message; I force-pushed a corrected commit to fix the error: https://github.com/vufind-org/vufind/commit/4b7fca9e033ba73702174be7b22c01f9426b585d -- apologies if this causes any confusion!
